tools/api-key-rotation/api_key_rotation_checker/main.py

```diff
@@ -0,0 +1,346 @@
+
+                # We need to convert
+                # our creation time for comparison
+                converted_creation_date = time_convert(key_object)
+
+                # Extract API Key ID for logging
+                key_id = key_object.key_id
+
+                # If our key is older than x days (default 90)
+                # based on our compare_dates function
+                # add api key to appropriate variable container
+                logging.info("Checking API key: %s creation date.. \n", key_id)
+                # Convert to JSON for logging
+                key_object_json = key_object.to_json()
+                if compare_dates(converted_creation_date, rotation_date):
+                    keys_needing_rotation.append(key_object_json)
+                else:
+                    keys_not_needing_rotation.append(key_object_json)
+
+    # Format our API keys
+    keys_needing_rotation = "\n".join(keys_needing_rotation)
+    keys_not_needing_rotation = "\n".join(keys_not_needing_rotation)
+
+    # Once analysis is complete for all keys,
+    # log results
+    if keys_needing_rotation:
+        logging.warning(
+            "Found API keys older than %s days. Please rotate: \n%s \n", rotation_period, keys_needing_rotation) # pylint: disable = line-too-long
+
+    if keys_not_needing_rotation:
+        logging.info(
+            "The following API key(s) are not older than %s days: \n%s", rotation_period, keys_not_needing_rotation) # pylint: disable = line-too-long
+
+
+def create_token():
+    """
+    As of March 2021, Google has released a SDK for API keys
+    but the SDK is in a private alpha and not accessible for
+    public usage. You can view the SDK here:
+    https://cloud.google.com/sdk/gcloud/reference/alpha/services/api-keys
+
+    Therefore, we must use the requests library and make API calls
+    with a bearer token until the SDK is usable by the public.
+    """
+    access_token = subprocess.run(
+        "gcloud auth print-access-token",
+        shell=True,
+        check=True,
+        stdout=subprocess.PIPE,
+        universal_newlines=True)
+    token = access_token.stdout
+
+    return token
+
+
+def compare_dates(converted_creation_date, rotation_date):
+    """
+    Compares createTime date to x (default 90) days ago.
+
+    Args:
+
+    converted_creation_date - The datatime formatted creation date of our API key.
+    rotation_date - datetime formatted "rotation_period" days ago (default 90).
+                    Example: 2020-09-18 13:38:52.943663
+    """
+
+    # If the createTime value for our key
+    # is over x days (default 90)
+    # Return true to key_analysis function
+    if converted_creation_date < rotation_date:
+        return True
+    else:
+        return False
+
+
+def create_service():
+    """
+    Creates the GCP Cloud Resource Service.
+    """
+    return googleapiclient.discovery.build(
+        "cloudresourcemanager",
+        "v1",
+        cache=MemoryCache())
```

Comment:
The cache is an optional parameter. Why you need it?

Reply:
> The cache is an optional parameter. Why you need it?

This cache helps with noisy logging from Cloud Functions. Without it there are numerous log entries generated due to issues with Google's usage of outdated packages.